homeassistant/components/homekit/type_fans.py

```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
                   ATTR_OSCILLATING: oscillating}
         self.call_service(DOMAIN, SERVICE_OSCILLATE, params, oscillating)
 
+    def set_speed_list(self, speed_list):
+        """Set and validate the speed list from the fan."""
+        if self.speed_list == speed_list:
+            return
+        if not speed_list:
+            self.speed_list = []
+            return
+        if SPEED_OFF != speed_list[0]:
+            _LOGGER.warning(
+                "%s: %s (%s) does not contain the speed setting "
+                "%s as its first element. "
+                "Assuming that %s is equivalent to 'off'.""",
+                self.entity_id, ATTR_SPEED_LIST, speed_list,
+                SPEED_OFF, speed_list[0])
+        self.speed_list = speed_list
+
+    def speed_from_percentage(self, percentage):
+        """Convert a value from 0-100 to a value in the speed list."""
+        if not self.speed_list:
+            return None
```

Comment:
If we pass the `speed_list_type` to those methods, we should be able to do direct mapping. No need for other checks and calculations. The mapping can be though a dictionary.

Reply:
Agreed